### DISCUSSION TITLE
Prevent query re-fetching on component remount

### DISCUSSION THREAD

**Discussion by antonmelbardis** [TanStack/query#616] [General]
Hi, 

I was hoping someone could help me with the following issue: I have an inactive query, which is a part of an unmounted component (that was previously mounted). Then, once the component is remounted the query is re-fetched, rather than being taken from the cache. Is there any way to prevent this from happening?

Thanks!

**Comment by tannerlinsley** [comment]
Please post a codesandbox or some code please. To help effectively, you should alway submit something more than a description :)

**Reply by antonmelbardis** [reply]
Hi Tanner, 

Apologies for not providing enough information in my first post. 

Please have a look at the code sample: https://codesandbox.io/s/react-query-refetching-on-mount-o05f1

This is a similar situation to what we are experiencing in our project.

**Reply by tannerlinsley** [reply]
So, right now, when a query gets mounted it is refetched in the background. It still uses the data in the cache even from when it was inactive, so you should at least see the old data while it fetches in the background. If you don't want a query instance to refetch when it mounts, use the `refetchOnMount: false` option. https://codesandbox.io/s/react-query-refetching-on-mount-nif07?file=/src/App.js

**Reply by GeospatialMax** [reply]
Hi @tannerlinsley , thanks for you reply. We had tried the `refetchOnMount` option already (directly on the query as well as on `ReactQueryConfigProvider`). It didn't make a difference unfortunately.
I was drilling a bit into a source code of react-query, looking for occurences of `refetchOnMount` and found [this line](https://github.com/tannerlinsley/react-query/blob/d7048d1c35efae5826988448055033fbc59d4c56/src/queryCache.js#L448) in queryCache.js

`(query.config.refetchOnMount || query.instances.length === 1)`

So basically if there's one instance of the query only it'll ignore the `refetchOnMount` option. So either this `or` statement would have to be removed to make this work or we would have to add another option like `alwaysRefetchOnMount` (which would ignore that there's only one instance).

I've changed to CodeSandbox to include a static second instance of `useQuery` for component 1: https://codesandbox.io/s/react-query-refetching-on-mount-o05f1?file=/src/App.js . You'll see in the console that it'll refetch the query for component 2 every time, but won't do it for component 1. But this would obviously be a strange workaround :-)

**Reply by tannerlinsley** [reply]
Hmmm... yeah, I think we might need to fix this. I'll look deeper into this to find out how to fix it properly.

**Reply by GeospatialMax** [reply]
@tannerlinsley We were wondering if you have any update on this? We would happy to try a fix ourselves and provide a Pull Request in line with my suggestion above for this, if that would help?

**Reply by tannerlinsley** [reply]
Yeah, sorry for the delay. A WIP PR would at least get us started in the right direction :)

**Reply by GeospatialMax** [reply]
@tannerlinsley Not a problem. Have now opened a draft PR here to get us started: https://github.com/tannerlinsley/react-query/pull/789.

**Reply by sid68v** [reply]
Umm.. any update on this guys ? :/

**Reply by lorstenoplo** [reply]
@sid68v it's fixed whats the problem, you just have to create a new queryCache and pass it to the client

**Reply by spiropoulos94** [reply]
> @sid68v it's fixed whats the problem, you just have to create a new queryCache and pass it to the client

Can you please provide an example? Thank you

**Reply by TkDodo** [reply]
I am not sure what the issue here is anymore, this is a very old discussion. Please create a minimal codesandbox example that shows what you think is not working as expected, then I can take a look.

**Reply by lorstenoplo** [reply]
@sid68v  For the site : https://goloop.vercel.app
For the code : https://github.com/lorstenoplo/eccom-next

You just have to initialise queryCache and provide it to the client options.

**Reply by TkDodo** [reply]
well, to be honest, I'm not gonna investigate a repo without knowing what the issue is. If you don't want to provide a minimal example, then at least _I_ am not going to put up more of my time, sorry.

**Comment by spiropoulos94** [comment]
Hello gentlemen, I would like to present you my solution for the query refetching on re-mount. I guess that everytime the component remounts, the queryKey is regenerated, therefore leading to useQuery being constantly re-updated even though its value is the same. What worked for me as a workaround is passing the queryKey from a parent component's state.  This way, I can navigate through my app while useQuery refteches only when the "globalState.text" is changed.

To sum up: 

1) Component mounts
2)Receives query key from parent as props (and mutates it  to a new one if it wants to )
3)Query key is the same as previous mount
4)Therefore the useQuery is exactly the same query as on previous mount.
5)useQuery does not refetch.

`

    const text = globalState.text

    const fetchDrinks = async (text) => {
        const key = text['queryKey'][0]

        const response = await fetch(`https://www.thecocktaildb.com/api/json/v1/1/search.php?s=${ key!== "" && key}`);
        const json = await response.json();
        return json;
    }

    const {isLoading, error, data} = useQuery(globalState.text, fetchDrinks, {
        refetchOnMount: false
    });`

**Reply by lorstenoplo** [reply]
Thanks, but I could not implement it. Can you help me with a repo with that implemented?

**Reply by spiropoulos94** [reply]
It is pretty simple actually, the steps are the following:

1. Decide on a query key name e.g "FetchingPeopleData"
2. Assign the query key name to a function that does the fetching with useQuery.
3. Make sure that the query name is passed to you component either by a parent component or by global state.

Take in mind that if you declare the queryKey like that _inside the component:_

`const queryKey = "Name"`

then the queryKey will be re-declared on every re-mount, forcing the useQuery to re-run every time.

So try to declare it on a **parent component or global state** and pass it down as a property, a property that is the same through every component mount.

**Comment by philals** [comment]
Anyone coming back to this with similar issue:

My `new QueryClient()` was inside my router. Every time the route changed, we got a new QueryClient - this caused many refetches.

**Reply by DanLowo** [reply]
This works for me. Thank you

**Comment by bruno-barros** [comment]
Just giving my two cents. I'v got myself on this anoing behavior. Every tab clicked making react-query fetch again.
As I'm using Nextjs. It runs on the server before runs on the client side. This behavior was causing the QueryClient to be recreate every time.
This did the job:
```
const queryClient = useRef(new QueryClient({
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
        refetchOnMount: false,
        cacheTime: 1 * 60 * 60 * 1000,
        staleTime: 1 * 60 * 60 * 1000,
        retry: 1
      },
    },
  }))
```
useRef ensure it runs only on the browser

**Reply by TkDodo** [reply]
relevant read: https://tkdodo.eu/blog/use-state-for-one-time-initializations

**Comment by IzzaSuni** [comment]
just found the solution for the exact same problem. it's on its query options check if data is equal between old and new data then set `refetchOnMount` to false, here's how I done it.

```
function useGetData(id){
  return useQuery(
    [id],
    () => fetcher(id),
    {
      enabled: !!id,
      refetchOnMount :false
 
      },
    }
  );
}
```

**Comment by jm-olo** [comment]
> Just giving my two cents. I'v got myself on this anoing behavior. Every tab clicked making react-query fetch again. As I'm using Nextjs. It runs on the server before runs on the client side. This behavior was causing the QueryClient to be recreate every time. This did the job:
> 
> ```
> const queryClient = useRef(new QueryClient({
>     defaultOptions: {
>       queries: {
>         refetchOnWindowFocus: false,
>         refetchOnMount: false,
>         cacheTime: 1 * 60 * 60 * 1000,
>         staleTime: 1 * 60 * 60 * 1000,
>         retry: 1
>       },
>     },
>   }))
> ```
> 
> useRef ensure it runs only on the browser

For anyone else that still encounters this issue.

A version of the `useRef` idea worked for me. I was experiencing this same refetching issue and realized that my app defined `const queryClient = new QueryClient({...})` inside of my top level component (`app.jsx`) which was creating a new instance of `queryClient` on every rerender. Simply moving `const queryClient = new QueryClient({...})` to outside my `app.jsx` solved the problem. No need for using `useRef`

**Comment by FraternalFiber** [comment]
My solution to this problem is to check if data is already in cache using `queryClient.getQueryData` and return it, otherwise fetch data.

```typescript
export const useData = (searchParams: TSearchParams) => {
  const queryClient = useQueryClient()

  return useQuery<TDataResult[]>({
    queryKey: ['searchResults', searchParams],
    queryFn: () => getData(searchParams, queryClient),
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
    staleTime: Infinity, 
    gcTime: Infinity,
  })
}
```

```typescript
export const getData = async (searchParams: TSearchParams, queryClient: QueryClient) => {
  const url = "some url"

  // If search results are in cache, return them to prevent a refetch
  const cachedData = queryClient.getQueryData(['searchResults', searchParams])

  if (cachedData) {
    return cachedData
  }

  return fetch(url)
      .then(res => res.json())
      .then(res => {
        return res.results
      })
      .catch((error) => {
        throw ('Error: ' + error)
      })
}
```

That way data is fetched only once. If a component is unmounted and mounted again, the service returns data from cache and doesn't make a refetch. Keep in mind that queryKey must be exactly the same to achieve this result.

**Reply by TkDodo** [reply]
that's literally what `staleTime:Infinity` does ...

**Reply by FraternalFiber** [reply]
But it doesn't solve my problem. I already have the property `staleTime` set to _Infinity_, but it still refetch in background after returning the data from cache. Is it a bug or am I doing something wrong?

**Reply by TkDodo** [reply]
I'd need to see a reproduction, likely something wrong on your end. you don't even need to turn off the `refetchOn...` flags because they default to refetching stale content only. An unstable query client is often the cause: https://tkdodo.eu/blog/react-query-fa-qs#2-the-queryclient-is-not-stable

**Comment by gabo71096** [comment]
In my case I was updating state inside the `queryFn`, causing to refetch, because my app starts from the root path at a `company` selection, but the user might start from a path like `companies/1` so I'd want to fetch the company first and then the rest.
In the first case, the `company_id` comes from the `company` selection (stored in state), in the second case, `company_id` comes from the path param, then fetching the company and storing it in state.
So to fix it, I created a constant:
`const c_id = company_id ? parseInt(company_id) : options.company.id;`
Here company_id comes from path param, and options.company.id is pulling it from state.
Then I used the `c_id` in the queryKey option in order to cache the fetching correctly. Before I was using `options.company.id` as queryKey, but because it was getting updated, it was triggering a refetch.

Here's the full example:

```
  const { company_id } = useParams();
  const { options, setOptions } = useContext(OptionsContext)!;
  const c_id = company_id ? parseInt(company_id) : options.company.id;
  const { data, error, isPending, isError } = useQuery({
    queryKey: ["serviceTypes", c_id],
    queryFn: async () => {
      if (company_id) {
        const company = await readCompanyCompaniesCompanyIdGet({ path: { company_id: c_id } });
        setOptions((prev) => ({ ...prev, company: { id: company.data!.id, name: company.data!.name } }));
      }

      const r = await readServiceTypesForCompanyCompaniesCompanyIdServiceTypesGet({
        path: { company_id: c_id },
      });

      return r.data;
    },
    staleTime: fiveMinutes,
  });
```

Hope it helps someone!
